### ISSUE TITLE
Missing CS8602 for calculated property

### ISSUE THREAD

**Issue by eugene-rw** [dotnet/roslyn#74057]
**Version Used**:8.0.302, 8.0.101

**Steps to Reproduce**:

call dotnet build for code

`using System;
using System.Collections.Generic;

#nullable enable

class ValueProvider
{
    private int _counter;

    public List<double>? NumberList
    {
        get
        {
            if (_counter++ == 0)
                return new List<double>();

            return null;
        }
    }
}

class Program
{
    static void Main()
    {
        var provider = new ValueProvider();

        if (provider.NumberList != null)
        {
            var a = provider.NumberList;

            if (a.Count == 0)
            {
                Console.WriteLine("Success");
            }
        }
    }
}
`
**Expected Behavior**:
Warning CS8602

**Actual Behavior**:
No warning

**Comment by eugene-rw** [comment]
It seems, compiler incorrectly use condition "provider.NumberList != null" as proof of non-nullability of `a`

**Comment by huoyaoyuan** [comment]
> It seems, compiler incorrectly use condition "provider.NumberList != null" as proof of non-nullability of `a`

It is the contract of property. Nullability analysis and many other analyzers doesn't expect the get result of property to change randomly. There are also other cases that tooling can invoke the getter of properties.

**Comment by eugene-rw** [comment]
@huoyaoyuan 
I would agree, if it was simple variable. But nullability analysis can see that the property is function indeed, and doesn't relay on it constancy. It can be make as configurable though.

**Comment by huoyaoyuan** [comment]
> But nullability analysis can see that the property is function indeed

Nullability analysis and potentially all language features doesn't see through method body of other methods. It only check for the signatures with attributes applied.

It's also impractical to understand what can happen in a method in detail.

> I would agree, if it was simple variable.

It's the contract of properties to act similar to simple variables. Many of the tooling including debugger follow this contract.

**Comment by eugene-rw** [comment]
> It's also impractical to understand what can happen in a method in detail.
Yes, of course.

I supposed, that there is a simple method to distinguish cases when property has simple get accessor with hidden backing field and when property calls function. But PropertyInfo has no info for that, it seems, that it can be compiler internals only.

**Comment by CyrusNajmabadi** [comment]
Closing as by design. Only the signatures of members are examined to determine their impact on NRT analysis.
